src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutesV2Spec.scala

```diff
@@ -57,45 +60,933 @@ class ResourceRoutesV2Spec extends FlatSpec with Matchers with TestSupport with
+    }
+  }
+
+  it should "404 when resource or resourceType does not exist" in {
+    val resourceType = ResourceType(
+      ResourceTypeName("rt"),
+      Set(SamResourceActionPatterns.readPolicies),
+      Set(ResourceRole(ResourceRoleName("owner"), Set(SamResourceActions.readPolicies))),
+      ResourceRoleName("owner")
+    )
+    val samRoutes = TestSamRoutes(Map(resourceType.name -> resourceType, resourceTypeAdmin.name -> resourceTypeAdmin))
+
+    samRoutes.resourceService.initResourceTypes().unsafeRunSync()
+
+    val resourceId = ResourceId("foo")
+    runAndWait(samRoutes.resourceService.createResource(resourceType, resourceId, samRoutes.userInfo, samRequestContext))
+
+    Get(s"/api/resources/v2/fakeResourceTypeName/${resourceId.value}/allUsers") ~> samRoutes.route ~> check {
+      status shouldEqual StatusCodes.NotFound
+    }
+
+    Get(s"/api/resources/v2/${resourceType.name}/fakeResourceId/allUsers") ~> samRoutes.route ~> check {
+      status shouldEqual StatusCodes.NotFound
+    }
+  }
+
+  it should "404 when user is not in any of the policies on the resource" in {
+    val resourceType = ResourceType(
+      ResourceTypeName("rt"),
+      Set(SamResourceActionPatterns.readPolicies),
+      Set(ResourceRole(ResourceRoleName("owner"), Set(SamResourceActions.readPolicies))),
+      ResourceRoleName("owner")
+    )
+    val samRoutes = TestSamRoutes(Map(resourceType.name -> resourceType, resourceTypeAdmin.name -> resourceTypeAdmin))
+
+    samRoutes.resourceService.initResourceTypes().unsafeRunSync()
+
+    val resourceId = ResourceId("foo")
+    runAndWait(samRoutes.resourceService.createResource(resourceType, resourceId, samRoutes.userInfo, samRequestContext))
+
+    val user = samRoutes.directoryDAO.loadUser(samRoutes.userInfo.userId, samRequestContext).unsafeRunSync().get
+    val userIdInfo = UserIdInfo(user.id, user.email, user.googleSubjectId)
+
+    Get(s"/api/resources/v2/${resourceType.name}/${resourceId.value}/allUsers") ~> samRoutes.route ~> check {
+      status shouldEqual StatusCodes.OK
+      responseAs[Set[UserIdInfo]].map(_.userSubjectId) shouldEqual Set(userIdInfo.userSubjectId)
+    }
+
+    val otherUserSamRoutes = TestSamRoutes(Map(resourceType.name -> resourceType), UserInfo(OAuth2BearerToken("accessToken"), WorkbenchUserId("user2"), WorkbenchEmail("user2@example.com"), 0))
+
+    Get(s"/api/resources/v2/${resourceType.name}/${resourceId.value}/allUsers") ~> otherUserSamRoutes.route ~> check {
+      status shouldEqual StatusCodes.NotFound
+    }
+  }
+
   private def mockPermissionsForResource(samRoutes: SamRoutes,
                                          resource: FullyQualifiedResourceId,
-                                         actionsOnResource: Set[ResourceAction] = Set.empty,
-                                         missingActionsOnResource: Set[ResourceAction] = Set.empty,
-                                         accessToResource: Boolean = true): Unit = {
```

Comment:
I simplified this to only need `actionsOnResource` because otherwise the mocked functions can be out of sync. Missing actions are missing if they are not in `actionsOnResource`. Access to resource is true if `actionsOnResource` is non-empty.